src/cpp/session/modules/connections/ConnectionsIndexer.cpp

```diff
@@ -0,0 +1,321 @@
+/*
+ * SessionConnectionsIndexer.cpp
+ *
+ * Copyright (C) 2009-16 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#include <core/Macros.hpp>
+#include <core/Algorithm.hpp>
+#include <core/Debug.hpp>
+#include <core/Error.hpp>
+#include <core/Exec.hpp>
+#include <core/FilePath.hpp>
+#include <core/FileSerializer.hpp>
+#include <core/text/DcfParser.hpp>
+
+#include <boost/regex.hpp>
+#include <boost/foreach.hpp>
+#include <boost/bind.hpp>
+#include <boost/range/adaptor/map.hpp>
+#include <boost/system/error_code.hpp>
+
+#include <r/RSexp.hpp>
+#include <r/RRoutines.hpp>
+#include <r/RExec.hpp>
+#include <r/RJson.hpp>
+#include <r/session/RSessionUtils.hpp>
+
+#include <session/projects/SessionProjects.hpp>
+#include <session/SessionModuleContext.hpp>
+#include <session/SessionPackageProvidedExtension.hpp>
+
+using namespace rstudio::core;
+
+namespace rstudio {
+namespace session {
+namespace modules { 
+namespace connections {
+
+
+namespace {
+
+bool isDevtoolsLoadAllActive()
+{
+   std::vector<std::string> search;
+   Error error = r::exec::RFunction("search").call(&search);
+   if (error)
+      return false;
+   
+   return std::find(search.begin(), search.end(), "devtools_shims") != search.end();
+}
+
+class ConnectionsIndexEntry
+{
+public:
+   
+   ConnectionsIndexEntry() {}
+   
+   ConnectionsIndexEntry(const std::string& name,
+                         const std::string& package,
+                         const std::string& shinyapp)
+      : name_(name), package_(package), shinyapp_(shinyapp)
+   {
+   }
+   
+   const std::string& getName() const { return name_; }
+   const std::string& getPackage() const { return package_; }
+   const std::string& getShinyApp() const { return shinyapp_; }
+
+   json::Object toJson() const
+   {
+      json::Object object;
+      
+      object["name"] = name_;
+      object["package"] = package_;
+      object["shinyapp"] = shinyapp_;
+      
+      return object;
+   }
+   
+private:
+   std::string name_;
+   std::string package_;
+   std::string shinyapp_;
+};
+
+class ConnectionsRegistry : boost::noncopyable
+{
+public:
+
+   void add(const std::string& package, const ConnectionsIndexEntry& spec)
+   {
+      connections_[constructKey(package, spec.getName())] = spec;
+   }
+
+   void add(const std::string& pkgName,
+            std::map<std::string, std::string>& fields)
+   {  
+      add(pkgName, ConnectionsIndexEntry(
+         fields["Name"],
+         pkgName,
+         fields["ShinyApp"]));
```

Comment:
Never mind, `std::map` won't throw: https://www.sgi.com/tech/stl/Map.html - "If the map does not already contain such an object, operator[] inserts the default object data_type()". So no need to catch. However, this makes me think that it would be nice to add a `warning` for the package creators, added now.